[IV-23-24] Objetivo 4: Tests

marinajcs opened the pull request:
* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
- Cada commit referencia a un issue, que a su vez referencia a una HU
* [x]  ¿Se ha intentado cubrir con tests una parte de la lógica de negocio que se
   esté desarrollando?
- Sí, puesto que se ha seguido TDD, primero se han desarrollado los tests y a partir
de ahí, el código necesario para implementar la lógica de negocio
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
- El milestone 1 describe el objetivo, el código entregable y qué lo hace viable (pasar los test)
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
- Sí, se ha priorizado implementar la clase `reparto.ts` frente al resto para cumplir el M1
* [x] ¿Estoy leyendo todo lo que marco o simplemente marco lo que se me pone por
delante?
- Sí a lo primero
* [x] ¿Se han documentado las elecciones de biblioteca de aserciones y del
   de *test runner*?
- Sí, en el fichero `docs/herramientas-test.md`
* [x] ¿Te has asegurado que lo que mencionas como biblioteca de aserciones y
test runners lo son realmente y tienen la misma funcionalidad? ¿O estás poniendo
`unittest` como *test runner* y `pytest` como biblioteca de aserciones?
- Sí, cada ejemplo y opción considerada es lo que viene documentado
* [x] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
- Sí, se han establecido unos criterios previos a la consideración y elección
de herramientas
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
- Sí, se han explicado las metodologías seguidas en la documentación
* [x]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?
- Sí, los issues que hacen referencia a testing están asignados al M1
* [x] ¿Se han testeado también las excepciones?
- Sí, en los tests también se prueba el manejo de excepciones

marinajcs committed:
Elección de herramientas de tests, closes #23

Incluida la justificación de la elección de Jasmine como test runner/framework en el fichero correspondiente.

marinajcs committed:
Integración de test runner, closes #23

Se ha instalado y configurado el test runner de Jasmine con pnpm y grunt

marinajcs committed:
Cambio en la elección de herramientas testing, fixes #23

Tras investigar, he decido que Chai y Mocha se adecuan más a los requisitos de mi proyecto, y que en los últimos años Jasmine ha sufrido un decrecimiento en el número de usuarios

marinajcs committed:
Desarrollo de tests para asignación de tareas, closes #26

Siguiendo la metodología TDD, primero se diseñan los test sabiendo que van a fallar, antes de proceder a escribir las funciones

marinajcs committed:
Implementado método de asignación de tareas, closes #25

Al crear un método de reparto de tareas, se pasan 2 tests sobre 4

marinajcs committed:
Lanza excepción si array compis está vacío, fixes #25

Se gestionan excepciones, se pasan 3 tests sobre 4

marinajcs committed:
Arreglada ordenación de las tareas en el reparto, fixes #25

Se ordenan las tareas disponibles de mayor a menor puntuación, se pasan 4 tests sobre 4

marinajcs committed:
Actualizado README con la orden para los test, fixes #26

Se incluye la orden a utilizar para que ejecute los tests correctamente

marinajcs committed:
Mejoras en asignarTareas, fixes #25

Se modulariza más el código y ahora se incluye una actualización del objetivo en caso que alguno no pueda cumplirlo y ya se haya terminado el reparto. Se pasan 4 tests sobre 4

marinajcs committed:
Actualizado fichero de elección de herramientas testing, fixes #23

Pequeñas correcciones sintácticas y ajuste de saltos de línea markdown

marinajcs committed:
Validación de datos previa a reparto, closes #28

Siguiendo la metodología TDD, se desarrollan tests para la validación y manejo de excepciones de las entidades de compañero y tarea

marinajcs commented:
Listo para revisión @JJ

JJ commented:
¿No prefieres esperar a que alguien más te lo revise?

marinajcs commented:
> ¿No prefieres esperar a que alguien más te lo revise?

Me parece bien, me espero unos días más 👍

PabloBarTo commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,112 @@
+del código de prueba. Su propio conjunto de Matchers para realizar aserciones de manera clara y expresiva.
+
+
+La elección final entre las aserciones anteriores han sido las de **Chai**, debido a los siguientes
```
typo han

marinajcs replied:
Corregido en https://github.com/marinajcs/asignacionTareas/pull/24/commits/985c43f338dcf28332d5c90a622c36e3ead5c4aa

PabloBarTo commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,112 @@
+Una vez escogidas las aserciones de Chai, se pueden utilizar un conjunto de herramientas que se
+integren bien con este entorno. Se han considerado los siguientes frameworks de testing:
+
+* **Jest**: como se ha mencionado anteriormente, es un marco de prubea conocido por su configuración
```
typo prubea

marinajcs replied:
Corregido en https://github.com/marinajcs/asignacionTareas/pull/24/commits/985c43f338dcf28332d5c90a622c36e3ead5c4aa

PabloBarTo commented on lib/reparto.ts:
```diff
@@ -3,35 +3,114 @@ import { Compi } from './compi';
 
-        //Debería empezar a asignar por el compi que tenga menos horas disponibles,
-        //para poder tener todas las combinaciones posibles para alcanzar el goal
+    getGoal(): number {
```
En TypeScript una mejor práctica sería definir el getter como "get goal()". Te dejo un enlace donde hay un ejemplo [getters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get)

marinajcs replied:
Corregido en https://github.com/marinajcs/asignacionTareas/pull/24/commits/a725c1c2ddcdd5ef98d6bd6922d5b768444adb39

PabloBarTo reviewed (changes requested):
Buenas! Te dejo aquí unos cambios pero en general veo todo bastante bien 😁

marinajcs committed:
Uso de get para getters, closes #29

Se adoptan mejores prácticas para la definición de getters en TypeScript

marinajcs committed:
Correciones ortográficas, fixes #23

Se corrigen pequeñas faltas de ortografía en el fichero de documentación de las herramientas de test

marinajcs commented:
> Buenas! Te dejo aquí unos cambios pero en general veo todo bastante bien 😁

Gracias por la revisión 😃
Ya he corregido lo que me has sugerido

marinajcs committed:
Corregida orden check, fixes #22

Se actualiza la tarea de comprobación de sintaxis en el fichero de configuración de Grunt, con el directorio de tests correcto

PabloBarTo reviewed (approved):
Por mi parte lo veo ya todo correcto👌

marinajcs commented:
Listo para revisión @JJ

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,112 @@
+TDD se centra principalmente en las pruebas unitarias, que son pruebas que verifican el
+comportamiento de unidades individuales de código (como funciones o métodos).
+
+TDD sigue un ciclo de desarrollo iterativo y repetitivo: *Red-Green-Refactor*:
```
¿Te parece que este texto (y lo que sigue) aporta algo a entender por qué has tomado una decisión sobre las bibliotecas usadas?

marinajcs replied:
Cierto, se ha eliminado la información extra de las metodologías y sintetizado la relevante.

Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/701a878821d1fa84a3f2a3adb244cd2b86fa1146

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,112 @@
+1. *Estándares y conformidad*: que cumpla los estándares relevantes de la industria
+de desarrollo de software, que sea compatible con el lenguaje (TypeScript) y el runtime
+(Node.js) elegidos.
+2. *Recomendaciones y comunidad*: una comunidad grande y activa casi siempre
```
Recuerda que todas las decisiones deben seguir criterios objetivos. Si tomas una decisión apoyada por este factor, debe haber algo que la apoye.

marinajcs replied:
Los criterios de comunidad, popularidad y mantenimiento ahora se respaldan por datos objetivos y cuantificables.

Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/efb16938f9ed311cbf26eb532cb7c3d1cbb06689

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,112 @@
+* **Chai**: es una biblioteca de aserciones flexible que se puede utilizar con diversos estilos
+de BDD. Ofrece las interfaces *expect*, *should*, y *assert*, lo que permite elegir el estilo
+que mejor se adapte a las preferencias de uno.
+* **Jest**: es un marco de prueba que incorpora su propia biblioteca de aserciones y se integra bien
```
Jest no es una biblioteca de aserciones, sino un marco completo de test. ¿Has mirado qué biblioteca de aserciones usa en realidad Jest? ¿Es posible que sea otra de las que ya has puesto?

marinajcs replied:
Es verdad, Jest es un framework todo en uno que solía depender de las aserciones de Jasmine y aunque ahora tiene las
suyas propias, esta siguen compartiendo similitudes con las de Jasmine.

Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/dd779952f9391c8c587469194cbbc599e15e02af

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,112 @@
+que mejor se adapte a las preferencias de uno.
+* **Jest**: es un marco de prueba que incorpora su propia biblioteca de aserciones y se integra bien
+con BDD. Utiliza una sintaxis sencilla y es conocido por su configuración fácil y rápida.
+* **Earl**: biblioteca de aserciones para JavaScript que busca proporcionar mensajes de error
```
Ya os comenté en clase que usárais herramientas como Snyk Advisor para filtrar (y quizás encontrar) bibliotecas aquí. ¿Has mirado qué puntuación tiene? ¿Crees que merece la pena insertarla?

marinajcs replied:
Se ha descartado Earl debido a su baja puntuación de salud.

Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/dd779952f9391c8c587469194cbbc599e15e02af

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,112 @@
+comprensión de las pruebas. Su sintaxis tipo BDD contribuye a una interpretación más clara del código.
+Ofrece flexibilidad con varios estilos de aserciones (should, expect, assert), permitiendo adoptar el estilo
+que mejor se alinee con cada usuario.
+2. *Recomendaciones y comunidad*: Chai es muy utilizado en la comunidad, con una base de usuarios activa
```
¿Como sabes lo que se usa?

marinajcs replied:
Se han consultado los datos de comunidad y popularidad recolectados en Snyk para las herramientas consideradas.

Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/efb16938f9ed311cbf26eb532cb7c3d1cbb06689

JJ reviewed (changes requested):
Por favor, trata de ajustar un poco más la documentación de la elección de herramientas a los criterios objetivos que se buscan, y a documentar sólo lo que sea estrictamente necesario para ello.
También trata de abrir un poco más el abanico de herramientas examinadas, y filtrar las que no tengan una salud adecuada.

marinajcs committed:
Elimina info. extra en metodologías, fixes #23

Se ha quitado información que no contribuye a la elección de herramientas y era de carácter meramente informativo acerca de las metodologías empleadas. Se sintetiza lo relevante

marinajcs committed:
Criterios objetivos: comunidad y mantenimiento, fixes #23

Se han modificado los criterios de recomendaciones y comunidad, añadiendo datos obetivos y cuantificables para proporcionar solidez a las elecciones tomadas

marinajcs committed:
Eliminados Jest y Earl en aserciones, fixes #23

Se han eliminado dado que no cumplen con los criterios de elección de aserciones. Jest porque no es una biblioteca de aserciones, sino un marco completo, y Earl porque tiene una puntuación baja dada su pequeña popularidad.

marinajcs committed:
Ampliar herramientas test consideradas, fixes #23

Se han añadido opciones en las bibliotecas de aserciones (las de AVA, Testdouble.js) y frameworks de testing (AVA, Tape), que cumplen con los criterios establecidos

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,102 @@
+* **Jasmine**: es un framework all-in-one que incluye tanto el marco de pruebas como la biblioteca de
+aserciones. Utiliza un estilo de desarrollo conducido BDD que enfatiza la legibilidad y la comprensión
+del código de prueba. Su propio conjunto de matchers para realizar aserciones de manera clara y expresiva.
+* **AVA**: es un marco de pruebas completo, pero también cuenta con sus propias funciones de aserción,
```
Por favor, incluid enlaces a las bibliotecas que estéis usando. `ava` tiene aserciones, pero es un marco de pruebas, como tú misma dices. ¿Has considerado la posiblidad de poner cada una en su apartado específico, de forma que sea más sencillo elegir alternativas?

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/480128a7f2b30f8dea56e29ab5c9a1f8a4e85c95

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,102 @@
+del código de prueba. Su propio conjunto de matchers para realizar aserciones de manera clara y expresiva.
+* **AVA**: es un marco de pruebas completo, pero también cuenta con sus propias funciones de aserción,
+que son simples y expresivas para facilitar la escritura de pruebas.
+* **Testdouble.js**: es una biblioteca de dobles (mocks, stubs y spies) versátil que puede emplearse en
```
... y por tanto no es una biblioteca de aserciones. Lo que mencionas son herramientas que se usan en tests, pero de ninguna manera una biblioteca de aserciones.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/21c41b21d27373913c964e002e077ff1c7bcbc51

JJ reviewed (changes requested):
En estos objetivos es esencial que, a través de las herramientas que mencionáis, mostréis que entendéis los conceptos que implementan. Una biblioteca de aserciones tiene una funcionalidad muy específica, y sería conveniente que entendieras bien qué es lo que hacen, y si las herramientas que listas corresponden *exclusivamente* a esa funcionalidad, antes de listarlas con el objeto de elegirlas o desecharlas.
Una regla del pulgar que se puede seguir es mirar el manual e intentar implementar alguna operación simple del tipo que se requiera: por ejemplo, si estás tratando de listar bibliotecas de aserciones, escoger una aserción de las que incluya y aplicarla a tu propio código. Si no es posible, es que simplemente no sigue la funcionalidad que se busca.
Y en todo caso te recuerdo la necesidad de enlazarlas, porque sólo por nombre es a vece simposible saber de qué se trata, y en todo caso hay que buscarlo.

JJ commented:
Adicionalmente, sería conveniente que examinaras bibliotecas de aserciones específicas de TypeScript, al menos las que tengan un cierto nivel de salud.

marinajcs committed:
Descartado Testdouble.js, fixes #23

Dado que no es una biblioteca de aserciones sino una de dobles, se elimina de las opciones consideradas

marinajcs committed:
Corrección de opciones de aserciones, fixes #23

En el apartado de aserciones, incluir SOLO aserciones, no frameworks que también tienen sus propias aserciones, para mayor simplicidad a la hora de realizar la selección de cada herramienta. Cada bibliotecas será enlazada a su correspondiente repositorio fuente.

marinajcs committed:
Añadidas bibliotecas específicas TS, fixes #23

Se mencionan algunas bibliotecas de aserciones concretas para TypeScript y/o JavaScript que han sido investigadas

marinajcs committed:
Mejoras en formato del documento, fixes #23

Pequeños ajustes de párrafos

marinajcs commented:
> En estos objetivos es esencial que, a través de las herramientas que mencionáis, mostréis que entendéis los conceptos que implementan. Una biblioteca de aserciones tiene una funcionalidad muy específica, y sería conveniente que entendieras bien qué es lo que hacen, y si las herramientas que listas corresponden _exclusivamente_ a esa funcionalidad, antes de listarlas con el objeto de elegirlas o desecharlas. Una regla del pulgar que se puede seguir es mirar el manual e intentar implementar alguna operación simple del tipo que se requiera: por ejemplo, si estás tratando de listar bibliotecas de aserciones, escoger una aserción de las que incluya y aplicarla a tu propio código. Si no es posible, es que simplemente no sigue la funcionalidad que se busca. Y en todo caso te recuerdo la necesidad de enlazarlas, porque sólo por nombre es a vece simposible saber de qué se trata, y en todo caso hay que buscarlo.

Entendido y gracias por el consejo, he consultado ejemplos de uso de las aserciones listadas en mi documento para
asegurarme de que podría implementarlas en mi desarrollo de tests, y así confirmar que realmente cumplen con la
funcionalidad que busco. También he incluido enlaces a las bibliotecas y frameworks que he investigado.

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,107 @@
+2. *Comunidad y popularidad*: una comunidad grande y activa; es decir con una gran cantidad
+de contribuidores y un elevado índice de uso entre los usuarios, casi siempre implica
+gran disponibilidad de recursos y ayuda de otros usuarios.
+3. *Mantenimiento*: se considerará que tiene un buen soporte en base a la frecuencia
```
¿No crees que esta esta relacionada con al anterior? O, en todo caso, si no está mantenida, ¿crees que tiene importancia que sea popular?

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/15adb7a81d9fd79aa3cfc01c31d79ecee7048c37

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,107 @@
+Para el desarrollo de tests, se van a seguir dos metodologías:
+
+1. TDD: determina cuándo se escriben los tests (antes de escribir el código de la app)
+2. BDD: determina el tipo de aserciones utilizadas (deben usar una sintaxis más cercana
```
Esto es en general correcto, pero como se entiende es que vas a escoger una biblioteca de aserciones que siga este formato. ¿Es así?

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/4fe57719c88fd7f4049ac0d4907061361e83856e

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,107 @@
+1. *Estándares y conformidad*: que cumpla los estándares relevantes de la industria
+de desarrollo de software, que sea compatible con el lenguaje (TypeScript) y el runtime
+(Node.js) elegidos.
+2. *Comunidad y popularidad*: una comunidad grande y activa; es decir con una gran cantidad
```
Aparte de lo que te digo abajo, si quieres examinar estos aspectos de forma objetiva tienes que usar herramientas como Snyk Advisor, que he comentado en clase repetidamente.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/15adb7a81d9fd79aa3cfc01c31d79ecee7048c37

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,107 @@
+para TS, segura en cuanto a tipos y fluida. Desafortunadamente, ha dejado de actualizarse y no cuenta con
+gran popularidad.
+
+Se han descartado bibliotecas de aserciones como Expect.js, Should.js, Unexpected, Earl y Power Assert
```
¿Y `hein`?

marinajcs replied:
Añadido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/431075356c2f771fe0eb88a87b38a5eed68cefbc

JJ commented on docs/herramientas-test.md:
```diff
@@ -0,0 +1,107 @@
+
+Se han descartado bibliotecas de aserciones como Expect.js, Should.js, Unexpected, Earl y Power Assert
+porque, aunque cumplían los requisitos de compatibilidad, su puntuación de "salud" era baja, debido a su
+descenso en popularidad o falta de mantenimiento. También otras bibliotecas como zod y typanion, ya que
```
¿Y crees por tanto que tiene importancia la comunidad si están bien mantenidas?

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/431075356c2f771fe0eb88a87b38a5eed68cefbc

JJ commented on iv.yaml:
```diff
@@ -3,3 +3,4 @@ entidad: lib/reparto.ts
 automatizar:
   fichero: Gruntfile.js
   orden: pnpm grunt
```
Hum, esto no lo había visto. Los task managers se tienen que ejecutar de forma independiente, y para ello tienes que instalarlo de forma global. Si lo usas así es que lo estás instalando en el `node_modules` local, y por tanto en realidad estás usando `pnpm` como task manager.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/2f898281f0272d840559bb53267de47aea1d8f2e

JJ commented on lib/reparto.ts:
```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
+        return totalPts/compis.length;
+    }
+
+    get goal(): number {
```
¿Este getter lo estás usando en algún lado?

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/06ae5129d39a56dd38b49e35ed00d78d18bec847

JJ commented on lib/reparto.ts:
```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
+    }
+
+    addTarea(t: Tarea): void {
+        if (t.duracionEstimada <= 0) {
```
¿Hay alguna posibilidad en la construcción de la tarea que esto suceda?
Estos errores debéis capturarlos en la construcción de un objeto; el objeto no debe construirse si está en un estado inadecuado o imposible.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/d8b59b4e34d79df2a1fec4dfd3f66b0bfb518f5c

JJ commented on lib/reparto.ts:
```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
+        if (t.duracionEstimada <= 0) {
+            throw new Error('La tarea debe tener una duración estimada válida');
+        } 
+        if (t.puntuacion <= 0) {
```
Igual, por eso se hacen objetos inmutables, para que la comprobación tenga que hacerse una sola vez.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/d8b59b4e34d79df2a1fec4dfd3f66b0bfb518f5c

JJ commented on lib/reparto.ts:
```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
+    }
+
+    removeTarea(t: Tarea): void {
+        this._tareasDisponibles.filter(tarea => tarea !== t);
```
Esto no hace lo que dice que hace: devuelve las tareas sin una de ellas, pero no la elimina. Conviene que uséis la palabra adecuada para cada una de las variables.
Por otro lado, también le estás pasando la tarea completa en vez de un identificador único de la tarea. Si la tarea necesita un identificador único, útil para almacenarla en la entidad, por ejemplo, eso debería estar en el modelo. Los bucles siempre cuestan tiempo, y los diccionarios ayudan siempre a recuperar la información de forma más eficiente.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/f573eb124515b915f7f2c01e3b41549a34bf87e7

JJ commented on lib/reparto.ts:
```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
 
-        //Debería empezar a asignar por el compi que tenga menos horas disponibles,
-        //para poder tener todas las combinaciones posibles para alcanzar el goal
+    removeCompi(c: Compi): void {
```
¿Hay alguna HU que requiera esto?

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/06ae5129d39a56dd38b49e35ed00d78d18bec847

JJ commented on lib/reparto.ts:
```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
-        //Debería empezar a asignar por el compi que tenga menos horas disponibles,
-        //para poder tener todas las combinaciones posibles para alcanzar el goal
+    removeCompi(c: Compi): void {
+        this._compis.filter(compi => compi !== c);
```
... y tampoco hace lo que dice.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/06ae5129d39a56dd38b49e35ed00d78d18bec847

JJ commented on lib/reparto.ts:
```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
+        const tareasAsignadas = asignaciones.get(c);
+
+        if (tareasAsignadas) {
+            tCompi = tareasAsignadas.map(tarea => tarea.duracionEstimada).reduce((total, t) => total + t, 0);
```
Buen uso de map/reduce.

JJ commented on lib/reparto.ts:
```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
+        this._compis.forEach(compi => asignaciones.set(compi, []));
+        // Ordenar de menor a mayor horas disponibles
+        const compisOrdenados = [...this._compis].sort((a,b) => a.horasDisponibles - b.horasDisponibles);
+        // Ordenar de mayor a menor puntuación
```
En código limpio no se usan comentarios. El uso de la orden `sort` ya indica claramente qué es lo que estás haciendo.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/da98b287d1c0cc0f86eb2ca43ffd4fceb859173c

JJ commented on lib/reparto.ts:
```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
+        let asignaciones = new Map<Compi, Tarea[]>();
+        this._compis.forEach(compi => asignaciones.set(compi, []));
+        // Ordenar de menor a mayor horas disponibles
+        const compisOrdenados = [...this._compis].sort((a,b) => a.horasDisponibles - b.horasDisponibles);
```
¿Por qué usas el spread (`...`) aquí? ¿No puedes aplicar sort directamente al array?

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/3f80f7a23f77ab4c771514375ee351c923cd11fe

JJ commented on lib/reparto.ts:
```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
+
+            if (compisDisponibles.length > 0) {
+                const compiAsignado = compisDisponibles[0];
+                asignaciones.get(compiAsignado)!.push(tarea);
```
¿Por qué la ! aquí? ¿Hay alguna posibilidad de que sea nulo? Si lo es, ¿qué alternativa hay?

JJ commented on lib/reparto.ts:
```diff
@@ -3,35 +3,113 @@ import { Compi } from './compi';
+
+        tareasOrdenadas.forEach(tarea => {
+            const compisDisponibles = compisOrdenados.filter(compi => {
+                const tiempoAsignado = this.calcularTiempoAsignado(compi, asignaciones);
```
Fíjate que necesitas calcular esto porque has optado por que this.compisOrdenados sea estática, y tenga las horas disponibles *al principio*. Si se recalculara cada vez que se hace una asignación y se actualizara con las horas disponibles *tras la asignación*, no necesitarías hacer esta operación.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/36d8fac0bf9c78b50c7b8148485b05b5901fd24d

JJ commented on package.json:
```diff
@@ -4,14 +4,24 @@
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "grunt test"
```
Esto va a funcionar bien esté grunt instalado localmente o globalmente. Pero `pnpm` crea un entorno en el que se ejecutan las tareas instaladas, y por lo tanto si funciona bien localmente no tiene porqué funcionar bien si simplemente usas `grunt test` (sin `pnpm` delante)

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/2de55fcaa61429db54b2fa0e30895163b951a049

JJ commented on package.json:
```diff
@@ -4,14 +4,24 @@
     "grunt": "^1.6.1",
     "grunt-exec": "^3.0.0",
+    "grunt-mocha-test": "^0.13.3",
+    "grunt-ts": "6.0.0-beta.22",
```
Y esto es así porque estás instalando el task manager como un módulo local; `grunt` no se usa desde `pnpm` (o no debería usarse) y por tanto no es una dependencia del proyecto.

marinajcs replied:
Pese a haber instalado Grunt globalmente (y comprobar que es correcta usando `--version` y verificando que su directorio binario global se encuentra el la variable `$PATH`), cuando intento ejecutarlo en el proyecto (`grunt`), lanza un mensaje diciendo que necesita instalarse localmente.
En la [guía oficial de instalación](https://gruntjs.com/installing-grunt), recomiendan instalar Grunt usando `(p)npm install grunt@VERSION --save-dev` para definirlo como devDependency, a pesar de no ser lo más adecuado como bien ha indicado. Ahora puedo usar la orden `grunt <tarea>` sin añadir `pnpm` delante, pero necesito incluirlo como dependencia del proyecto para que funcione.

Contemplado en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/2de55fcaa61429db54b2fa0e30895163b951a049

JJ replied:
Deberías abrir un issue con el error y tratar de probar diferentes formas de solucionarlo. El problema aquí es que si escoges un task manager que tienes que ejecutar desde el dependency manager, no estás ganando tanto en realidad. También el modo de instalación va a influir en como definas el contenedor en el objetivo siguiente. Así que conviene que resuelvas el problema ahora.

marinajcs replied:
Entendido, investigaré a ver si hay alguna forma de usar Grunt sin tenerlo como dependencia y si veo que no hay forma, cambiaré de task manager (probablemente a pnpm run).

JJ replied:
El problema más que grunt en sí pueden ser los plugins, que tendrás que instalar también globalmente.

JJ commented on package.json:
```diff
@@ -4,14 +4,24 @@
     "typescript": "^5.2.2"
+  },
+  "dependencies": {
+    "@npmcli/fs": "^3.1.0"
```
¿Esto es una dependencia de producción?

marinajcs replied:
Eliminada en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/2de55fcaa61429db54b2fa0e30895163b951a049

JJ commented on test/reparto.test.ts:
```diff
@@ -0,0 +1,122 @@
+const Laura = new Compi(3);
+const Fernando = new Compi(10);
+
+const tarea1: Tarea = {
```
Mejor qeu esto es usar un array, y mucho mejor un fichero externo que se pueda modificar fácilmente.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/b0bdbb0908791b9f1f00d7a089ae66a59d58792a

JJ commented on test/reparto.test.ts:
```diff
@@ -0,0 +1,122 @@
+import { Reparto, Tarea } from '../lib/reparto';
+import { Compi } from '../lib/compi';
+
+const Rebeca = new Compi(6);
```
DRY, usa un array y no repitas código ni definas más variables de las estrictamente necesarias.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/b0bdbb0908791b9f1f00d7a089ae66a59d58792a

JJ commented on test/reparto.test.ts:
```diff
@@ -0,0 +1,122 @@
+        }
+    });
+
+    it('debería lanzar una excepción si se le pasa alguna tarea con puntuación no válida', () => {
```
Buen testeo de excepciones.

JJ commented on test/reparto.test.ts:
```diff
@@ -0,0 +1,122 @@
+        const gestorTareas = new Reparto([Rebeca, Laura], [tarea1, tarea2, tarea5]);
+        const asignaciones = gestorTareas.asignarTareas();
+
+        expect(asignaciones.get(Rebeca)).to.not.include(tarea5);
```
Como pone en [los errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-4.md#sobre-los-tests):

> El error principal es testear que las funciones dan la salida esperada, es decir, aplicar la lógica de negocio y si uno espera que el resultado sea X, comprobar que es X. En realidad, no es así. Hay que comprobar que se cumple lo que pide la lógica de negocio. Si se pide que algo sea mínimo, o mejor, u óptimo, hay que comprobar que efectivamente el resultado es el mínimo, o mejor que otros, u óptimo; hay que tener una función que compruebe la "bondad" o lo que sea de la solución y mostrar que efectivamente es la mejor según ese criterio. Esto es repetir lo de arriba, pero hay que tenerlo en cuenta a la hora de escribir los tests.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/a92897090f957fac2e4bd8e0f30148fdf8fad414

JJ reviewed (changes requested):
Siento la cantidad de correcciones; algunas son "kudos", pero en todo caso estás más cerca de alcanzar el objetivo. El problema principal es el testeo de la lógica de negocio, que no puede corresponder a las salidas de una versión específica de la misma, porque se rompería en caso de que otra función alcance otras salidas, siempre cumpliendo la historia de usuario.

marinajcs committed:
Instalación global de Grunt, fixes #19

Se usa Grunt de forma independiente para la automatización de tareas en lugar de usar el entorno de pnpm

marinajcs committed:
Corrección de criterios, fixes #23

Se unifican los criterios de comunidad y mantenimiento, puesto que guardan relación entre sí. El mantenimiento es crucial

marinajcs committed:
Aserciones son de tipo BDD, fixes #23

Aclaración de que la biblioteca de aserciones escogida debe ser compatible con el estilo BDD (como es en el caso de Chai)

marinajcs committed:
Alteración en las opciones de aserciones, fixes #23

Se ha mencionado hein, otra biblioteca específica para TS, y se ha aclarado que el mantenimiento prima sobre la comunidad/popularidad de una herramienta, aunque a menudo estén vinculadas entre sí

marinajcs committed:
Orden grunt sin pnpm, fixes #19

La orden para ejecutar tareas es grunt <tarea>, sin el pnpm delante. No obstante, de acuerdo con la documentación oficial de Grunt, Grunt y sus plugins deben definirse como devDependencies del package.json del proyecto, a pesar de haberse instalado Grunt globalmente

marinajcs committed:
Eliminar comentarios de código limpio, fixes #25

marinajcs committed:
Eliminar métodos prescindibles, fixes #25

Las funciones de get goal, add y remove Compi han sido eliminadas, ya que no las pide la HU

marinajcs committed:
Uso de arrays y fichero externo datos test, fixes #26

Se evita declarar muchas variables innecesarias y que repiten código, usando arrays y un fichero externo que almacena las tareas para mayor comodidad a la hora de manejar los datos del test

marinajcs committed:
ID de tarea y removeTarea, fixes #25

Se ha creado un ID de tarea para aumentar la eficiencia de recuperación de información, especialmente en el método removeTarea(), que se ha corregido para que realice la función que debería.

marinajcs committed:
Suprimido el spread, fixes #25

Se ha eliminado el spread (...) ya que no aportaba nada a la hora de realizar el sort

marinajcs committed:
Se recalculan las horasDisponibles tras asignación, fixes #25

Se elimina la función calcularTiempoAsignado y las llamadas a la misma, pues ahora se recalcula el vector de compisOrdenados y se actualizan las horasDisponibles tras cada asignación.

marinajcs committed:
Valida datos antes de construir objetos, fixes #25

La comprobación ahora se realiza una única vez antes de construir el objeto, ya que no deben declararse con datos no válidos.

marinajcs committed:
Tests no dependen de datos concretos, fixes #26

El éxito o fracaso de las pruebas ya no depende de los datos usados, pues se han implementado funciones a las que se le pasa cualquier tipo de asignación.

marinajcs commented:
Resulta que Grunt no está pensado para usarse globalmente, pues aunque `grunt-cli` se instala de forma global, cuando se ejecuta el comando `grunt`, lo que hace es buscar una instalación local de Grunt en el directorio y subdirectorios en el que se encuentra actualmente y usarla para cargar el `Gruntfile.js`.
He investigado foros de usuarios intentando ejecutarlo de forma global, y la única forma es usando enlaces simbólicos a una carpeta `node_modules` (donde se instalaría manualmente Grunt y todos los plugins usados) creada en la raíz del usuario, pero dudo que sea una buena práctica, además de que "colisiona" con el `node_modules` que ya tiene el proyecto. Sería conveniente cambiar a pnpm run o cualquier otro task manager, ¿no?

Me aseguraré de resolver este issue antes de empezar con el objetivo 5, pero le etiqueto igualmente por si debo corregir más cosas de los tests y entidades. @JJ

JJ commented:
> Resulta que Grunt no está pensado para usarse globalmente, pues aunque `grunt-cli` se instala de forma global, cuando se ejecuta el comando `grunt`, lo que hace es buscar una instalación local de Grunt en el directorio y subdirectorios en el que se encuentra actualmente y usarla para cargar el `Gruntfile.js`. He investigado foros de usuarios intentando ejecutarlo de forma global, y la única forma es usando enlaces simbólicos a una carpeta `node_modules` (donde se instalaría manualmente Grunt y todos los plugins usados) creada en la raíz del usuario, pero dudo que sea una buena práctica, además de que "colisiona" con el `node_modules` que ya tiene el proyecto. Sería conveniente cambiar a pnpm run o cualquier otro task manager, ¿no?

Gracias por investigar esto, no lo sabía. Pero realmente sí, parece más simple, sobre todo a este nivel. Es posible que si tienes workflows más complejos, con diferentes fases de conversión de los fuentes y demás, convenga usarlo, pero el problema es que ya los bundlers como `webpack` y demás son capaces de manejar todas esas fases, y de hecho pueden comportarse como mini-gestores de tareas (más build tools, pero pueden hacer este tipo de cosas).

Quizás sea mejor optar por lo más simple; esta investigación que has hecho es suficiente para superar el objetivo :+1: 

> 
> Me aseguraré de resolver este issue antes de empezar con el objetivo 5, pero le etiqueto igualmente por si debo corregir más cosas de los tests y entidades. @JJ

Pide revisión y lo veo; ya he visto que has resuelto todos los 24 issues :clap:, pero tendrás que poner la decisión de lo de arriba en la práctica. Si acaso resume lo que has comentado arriba como razón para *no* elegir `grunt`

marinajcs committed:
Cambiada elección task manager, closes #31

Se ha cambiado el task manager por los motivos explicados en el issue referenciado y la documentación de gestor-tareas.md.

JJ commented on Gruntfile.js:
```diff
@@ -2,11 +2,26 @@ module.exports = function (grunt) {
     grunt.initConfig({
         exec: {
-            typescript: 'tsc --noEmit lib/*.ts'
+            typescript: 'tsc --noEmit --target es6 lib/*.ts test/*.ts',
```
Pero aquí sigue grunt... No tiene importancia, puedes borrarlo antes de mergear.

marinajcs replied:
Eliminado en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/a6c29afbbc326bfdc941d6331067a5b8a7ec14c0

JJ commented on package.json:
```diff
@@ -4,14 +4,21 @@
     "grunt": "^1.6.1",
     "grunt-exec": "^3.0.0",
-    "typescript": "^5.2.2"
+    "grunt-mocha-test": "^0.13.3",
```
Asegúrate de desinstalarlo también de aquí.

marinajcs replied:
Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/a6c29afbbc326bfdc941d6331067a5b8a7ec14c0

JJ commented on package.json:
```diff
@@ -4,14 +4,21 @@
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": ""
```
¿Aquí no hay nada? ¿Es correcto?

marinajcs replied:
Faltaba añadir el comando de mocha

Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/a6c29afbbc326bfdc941d6331067a5b8a7ec14c0

JJ reviewed (changes requested):
¿Seguro que has hecho correctamente la configuración de `pnpm`?

marinajcs committed:
Configuración de pnpm run, fixes #31

Se ha cambiado la configuración para adoptar pnpm como task manager del proyecto

marinajcs committed:
Eliminado Gruntfile.js, fixes #31

Ya no sirve el fichero de configuración de tareas de Grunt, puesto que se va a usar de ahora en adelante pnpm run.

marinajcs commented:
> ¿Seguro que has hecho correctamente la configuración de `pnpm`?

Ahora sí, se me había olvidado modificar el `iv.yaml` y los scripts del `package.json`

marinajcs committed:
Correcciones en package.json y tsconfig, fixes #31

Para la correcta ejecución de las tareas check y test

JJ commented on package.json:
```diff
@@ -4,14 +4,19 @@
-    "grunt": "^1.6.1",
-    "grunt-exec": "^3.0.0",
-    "typescript": "^5.2.2"
+  "dependencies": {
```
¿Crees que estas dependencias son de producción o de desarrollo? Si antes typescript era de desarrollo, ¿por qué ha pasado ahora a producción?
En todo caso, debéis tener cuidado en no editar directamente este apartado, sino usar solo pnpm install (con `--save-dev`) y `pnpm uninstall`

marinajcs replied:
Cierto, se me olvidó indicar la opción `--save-dev` cuando las volví a instalar y se guardaron como dependencias de producción en lugar de de desarrollo, como estaban antes.

Corregido en: https://github.com/marinajcs/asignacionTareas/pull/24/commits/da4d5ec17148c06f71b93301690a9cf4b2704021

JJ reviewed (changes requested):
Me temo que entender y poner en práctica el concepto de dependencias de desarrollo y de producción es suficientemente importante como para no superar, esta vez, el objetivo. Por favor, revísalo y ya estaría.

marinajcs committed:
devDependencies corregidas, fixes #31

Se han desinstalado de forma limpia todas las dependencias de producción y se han instalado como dependencias de desarrollo

JJ reviewed (approved):
:clap:

marinajcs committed:
Comandos pnpm run actualizados, fixes #31

Se cambian las órdenes antiguas de grunt por las de pnpm run
